src/components/shared/chart/Viewport.js

```diff
@@ -88,422 +48,521 @@ require('./Viewport.css');
-      // is not fully layed out correctly yet.
-      this._setSize();
-      this._setSizeNextFrame();
-    }
+      render() {
+        const { chartProps, hasZoomedViaMousewheel } = this.props;
+
+        const {
+          containerWidth,
+          containerHeight,
+          viewportTop,
+          viewportBottom,
+          viewportLeft,
+          viewportRight,
+          isDragging,
+          isShiftScrollHintVisible,
+        } = this.state;
+
+        const viewportClassName = classNames({
+          chartViewport: true,
+          dragging: isDragging,
+        });
 
-    componentWillUnmount() {
-      window.removeEventListener('resize', this._setSizeNextFrame, false);
-      window.removeEventListener('mousemove', this._mouseMoveListener, true);
-      window.removeEventListener('mouseup', this._mouseUpListener, true);
-    }
+        const shiftScrollClassName = classNames({
+          chartViewportShiftScroll: true,
+          hidden: hasZoomedViaMousewheel || !isShiftScrollHintVisible,
+        });
+
+        const viewport: Viewport = {
+          containerWidth,
+          containerHeight,
+          viewportLeft,
+          viewportRight,
+          viewportTop,
+          viewportBottom,
+          isDragging,
+        };
 
-    render() {
-      const { hasZoomedViaMousewheel } = this.props;
-
-      const {
-        containerWidth,
-        containerHeight,
-        viewportTop,
-        viewportBottom,
-        viewportLeft,
-        viewportRight,
-        isDragging,
-        isShiftScrollHintVisible,
-      } = this.state;
-
-      const viewportClassName = classNames({
-        chartViewport: true,
-        dragging: isDragging,
-      });
-
-      const shiftScrollClassName = classNames({
-        chartViewportShiftScroll: true,
-        hidden: hasZoomedViaMousewheel || !isShiftScrollHintVisible,
-      });
-
-      return (
-        <div
-          className={viewportClassName}
-          onWheel={this._mouseWheelListener}
-          onMouseDown={this._mouseDownListener}
-          ref={this._takeContainerRef}
-        >
-          <WrappedComponent
-            {...this.props}
-            containerWidth={containerWidth}
-            containerHeight={containerHeight}
-            viewportLeft={viewportLeft}
-            viewportRight={viewportRight}
-            viewportTop={viewportTop}
-            viewportBottom={viewportBottom}
-            isDragging={isDragging}
-          />
-          <div className={shiftScrollClassName}>
-            Zoom Chart:
-            <kbd className="chartViewportShiftScrollKbd">Shift</kbd>
-            <kbd className="chartViewportShiftScrollKbd">Scroll</kbd>
+        return (
+          <div
+            className={viewportClassName}
+            onWheel={this._mouseWheelListener}
+            onMouseDown={this._mouseDownListener}
+            ref={this._takeContainerRef}
+          >
+            <ChartComponent {...chartProps} viewport={viewport} />
```

Comment:
here you can pass anything as `viewport` because when you ask Flow to infer the type, it looks how you use it and manage to match. In this case it infers it with whatever you put in viewport (I don't know how it works when you use it then though).

This works if I use the following type for `ChartComponent`: `ChartComponent: React.ComponentType<$Subtype<{ +viewport: Viewport }>>` although there may be a simpler (without Subtype) or more complete (with charProps) solution.
And then you don't have to type "viewport" strongly above, Flow will properly check that you pass the correct properties.

Reply:
Ok, modifying the this. Would you mind in the future to put longer code snippets in comments in triple backticks for readability? It was really hard to read the `ChartComponent:` stuff.